Build HttpStress and SslStress with live-built runtime using current TFM

antonfirsov opened the pull request:
As noted in https://github.com/dotnet/runtime/issues/60769#issuecomment-951114180, the assumption that the live-built runtime of version `(X+1).0` can be a drop-in replacement in an SDK of version `X.0` is sometimes wrong, and it currently breaks our stress builds.

This PR changes both local (non-containerized) and containerized stress builds to build against the live-built runtime with the help of [`targetingpacks.targets`](https://github.com/dotnet/runtime/blob/main/eng/targetingpacks.targets).

Some tricky parts:
- The docker builds no longer replace the shared dotnet runtime. Instead, we copy the live built runtime dll-s and ref dlls under `/live-runtime-artifacts` and reference that location from properties required by `targetingpacks.targets`
- We also copy testhost to `/live-runtime-artifacts`. Tests are being executed by:
`TESTHOSTDIR/dotnet exec STRESSBINDIR/HttpStress.dll <args>` .
- Since I wanted to avoid using the entire repository as the container context, `targetingpacks.targets` also goes to `/live-runtime-artifacts`
- HttpStress needs compatible (7.0) `Microsoft.AspNetCore.App` bits. These are acquired by getting the latest daily SDK build with `dotnet-install.sh/ps1`
    - This mechanism replaces the outdated `privateAspNetCore` stuff implemented in #33860. Scripts args and dockerfiles dealing with that have been removed.
- To support local, non-containerized test execution helper scripts `build-local.ps1/sh` have been implemented. In case of HttpStress these scripts need to grab the daily SDK just like container scripts.
- Visual Studio experience is now broken for HttpStress, but it can be "fixed" by temporary local hacks if necessary. We may open a tracking issue to fix this.

Fixes #60769.

/cc @ManickaP @safern

antonfirsov committed:
refactor stress test builds

antonfirsov committed:
cleanup container scripts

antonfirsov committed:
delete LiveBuildTest

antonfirsov committed:
cleanup stress projects

antonfirsov committed:
remove unnecessarry newlines

antonfirsov committed:
cleanup dockerfiles

antonfirsov commented on src/libraries/System.Net.Http/tests/StressTests/HttpStress/StressServer.cs:
```diff
@@ -315,6 +317,13 @@ private static void MapRoutes(IEndpointRouteBuilder endpoints)
+        {
+            // For some reason, System.Security.Cryptography.Encoding.dll fails to resolve when being loaded on-demand by AspNetCore.
+            // Enforce early-loading to workaround this issue.
+            _ = new Oid();
```
Here is the error I get by removing this line:
https://dev.azure.com/dnceng/public/_build/results?buildId=1459792&view=logs&j=0da5d1d9-276d-5173-c4c4-9d4d4ed14fdb&t=6c56d998-158d-5232-283c-5140104799fc

ManickaP replied:
@wfurt any ideas? Should we file this as a separate issue? Or is this related to all the targeting pack magic around?

antonfirsov commented:
/azp run runtime-libraries stress-http

antonfirsov commented:
/azp run runtime-libraries stress-ssl

safern commented on eng/docker/libraries-sdk.linux.Dockerfile:
```diff
@@ -3,26 +3,45 @@ ARG BUILD_BASE_IMAGE=mcr.microsoft.com/dotnet-buildtools/prereqs:centos-7-f39df2
-ARG OS=Linux
+# Install 7.0 SDK:
+RUN wget https://dot.net/v1/dotnet-install.sh
+RUN bash ./dotnet-install.sh --channel 7.0.1xx --quality daily --install-dir /usr/share/dotnet
```
can we use the version instead so that this is not hardcoded and can be specified as an ARG? something like `--channel $VERSION.0.1xx

safern commented on src/libraries/System.Net.Http/tests/StressTests/HttpStress/Directory.Build.props:
```diff
@@ -1 +1,15 @@
+        <PackageRid Condition="$([MSBuild]::IsOSPlatform('WINDOWS'))">win-x64</PackageRid>
+        <PackageRid Condition="$([MSBuild]::IsOSPlatform('OSX'))">osx-x64</PackageRid>
+        
+        <TargetingPacksTargetsLocation Condition="'$(TargetingPacksTargetsLocation)' == ''">../../../../../../eng/targetingpacks.targets</TargetingPacksTargetsLocation>
```
There should be a `RepoRoot` property available?

safern replied:
We could get the root of the repo with something like: `$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildThisFileDirectory), global.json))`

antonfirsov commented:
/azp run runtime-libraries stress-http

antonfirsov committed:
better way to find reference locations in stress .props

also delete OSX <PackageRid>,
since OSX stress runs are practically unsupported

antonfirsov committed:
Dockerfiles: use $VERSION, delete $ARCH

antonfirsov commented:
/azp run runtime-libraries stress-http

antonfirsov committed:
fix windows.Dockerfile

antonfirsov committed:
fix Dockerfiles

antonfirsov committed:
fix linux.Dockerfile

antonfirsov commented:
/azp run runtime-libraries stress-http

antonfirsov commented:
/azp run runtime-libraries stress-ssl

antonfirsov committed:
fix linux linux.Dockerfile

antonfirsov commented:
/azp run runtime-libraries stress-http

antonfirsov commented:
/azp run runtime-libraries stress-http

antonfirsov commented:
/azp run runtime-libraries stress-http

antonfirsov committed:
disable HTTP/3 to see if the rest succeeds

antonfirsov committed:
Merge branch 'live-stress' of https://github.com/antonfirsov/runtime into live-stress

antonfirsov commented:
/azp run runtime-libraries stress-http

ManickaP commented:
Haven't done proper review yet, but I've noticed one small thing that slightly bothered me. Is there a reason why `Build-Local.ps1` filename is capitalized? All the other `sh` and `ps1` scripts in the directory are lower-cased.

antonfirsov commented:
I followed PS convention (`Copy-Item`, `Write-Host` etc), but you are right that all dotnet/runtime scripts are named in lowercase `do-something.ps1` style, so I will rename.

antonfirsov committed:
rename Build-Local.ps1 and extend README.md

antonfirsov committed:
re-enable HTTP3 with timeout

antonfirsov commented:
/azp run runtime-libraries stress-http

antonfirsov committed:
remove trailing whitespace

antonfirsov commented:
/azp run runtime-libraries stress-http

antonfirsov commented:
/azp run runtime-libraries stress-ssl

antonfirsov commented on eng/pipelines/libraries/stress/http.yml:
```diff
@@ -57,6 +57,7 @@ jobs:
       export HTTPSTRESS_CLIENT_ARGS="$HTTPSTRESS_CLIENT_ARGS -http 3.0"
       export HTTPSTRESS_SERVER_ARGS="$HTTPSTRESS_SERVER_ARGS -http 3.0"
       docker-compose up --abort-on-container-exit --no-color
+    timeoutInMinutes: 35 # In case the HTTP/3.0 run hangs, we timeout shortly after the expected 30 minute run
```
I'm adding this because I experienced mysterious HTTP/3.0 hangs on some runs:

https://dev.azure.com/dnceng/public/_build/results?buildId=1478084&view=logs&j=0da5d1d9-276d-5173-c4c4-9d4d4ed14fdb&t=1451f5f3-0108-5a08-5b92-e984b2a85bbd

ManickaP replied:
Do we know what happened? I see only blank screen (no output from the run)? Either way, we shouldn't hold this PR on it and investigate it separately.

antonfirsov replied:
It stopped logging after [this](https://dev.azure.com/dnceng/public/_build/results?buildId=1481385&view=logs&j=0da5d1d9-276d-5173-c4c4-9d4d4ed14fdb&t=1451f5f3-0108-5a08-5b92-e984b2a85bbd&l=62) line, then that log disappeared for some reason after the whole job becoming cancelled. Unfortunately I wasn't able to reproduce this behavior locally.

I would be surprised if this is caused by my changes, we haven't seen working stress runs on the CI for a while.

ManickaP commented on eng/docker/build-docker-sdk.ps1:
```diff
@@ -34,12 +29,35 @@ if ($buildWindowsContainers)
+  # Collect the following artifacts to folder, that will be used as build context for the container,
+  # so projects can build and test against the live-built runtime:
+  # 1. Reference assembly pack (microsoft.netcore.app.ref)
+  # 2. Runtime pack (microsoft.netcore.app.runtime.linux-x64)
```
```suggestion
  # 2. Runtime pack (microsoft.netcore.app.runtime.win-x64)
```
?

ManickaP commented on eng/docker/build-docker-sdk.ps1:
```diff
@@ -34,12 +29,35 @@ if ($buildWindowsContainers)
+      Remove-Item -Recurse -Force $dockerContext
+  }
+
+  Copy-Item -Recurse -Destination $dockerContext\microsoft.netcore.app.ref -Path $binArtifacts\microsoft.netcore.app.ref
```
NIT: can we swap order of -Path and -Destination? Or did you take the ordering here from somewhere else where we use it this way?
I was super confused since copy sans options is usually `source destination` argument order.

ManickaP commented on eng/docker/build-docker-sdk.ps1:
```diff
@@ -34,12 +29,35 @@ if ($buildWindowsContainers)
+  Copy-Item -Recurse -Destination $dockerContext\testhost -Path $binArtifacts\testhost
+  Copy-Item -Recurse -Destination $dockerContext\targetingpacks.targets -Path $REPO_ROOT_DIR\eng\targetingpacks.targets
+  
+  # In case of non-CI builds, testhost may already contain Microsoft.AspNetCore.App (see Build-Local.ps1 in HttpStress):
```
```suggestion
  # In case of non-CI builds, testhost may already contain Microsoft.AspNetCore.App (see build-local.ps1 in HttpStress):
```

ManickaP commented on eng/docker/libraries-sdk.linux.Dockerfile:
```diff
@@ -4,25 +4,47 @@ ARG SDK_BASE_IMAGE=mcr.microsoft.com/dotnet/nightly/sdk:6.0-bullseye-slim
+
+# Install latest daily SDK:
+RUN wget https://dot.net/v1/dotnet-install.sh
+RUN bash ./dotnet-install.sh --channel $_DOTNET_INSTALL_CHANNEL --quality daily --install-dir /usr/share/dotnet
```
Do we need `bash` here?

antonfirsov replied:
You need to precede the call with `chmod +x ./dotnet-install.sh` otherwise, which seems more complicated to me.

ManickaP commented on eng/docker/libraries-sdk.linux.Dockerfile:
```diff
@@ -4,25 +4,47 @@ ARG SDK_BASE_IMAGE=mcr.microsoft.com/dotnet/nightly/sdk:6.0-bullseye-slim
-ARG ARCH=x64
+ARG VERSION=7.0
 ARG CONFIGURATION=Release
+ENV _DOTNET_INSTALL_CHANNEL="$VERSION.1xx"
```
How often will we need to change this? Only with major versions (7.0 --> 8.0)? Or with every Preview/RC?

ManickaP commented on src/libraries/System.Net.Http/tests/StressTests/HttpStress/Directory.Build.props:
```diff
@@ -1 +1,17 @@
+        <!-- Stress projects have their own global.json, the directory above that also has it is the repository root. -->
+        <RepositoryRoot>$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildThisFileDirectory)../, global.json))/</RepositoryRoot>
+        
+        <TargetingPacksTargetsLocation Condition="'$(TargetingPacksTargetsLocation)' == ''">$(RepositoryRoot)eng/targetingpacks.targets</TargetingPacksTargetsLocation>
```
I will not pretend I understand all this targeting stuff, but I'm happy someone from our team does :rofl:

antonfirsov replied:
I would be also happy if someone explained the terminology properly here. From what I understand, a "targeting pack" allows you to target a specific runtime / tfm in your build, in this case the one you live-build locally. `targetingpacks.targets` somehow makes the globally installed SDK understand it, and allow to build against it, but it needs some ceremony to work.

ManickaP commented:
I can build the `HttpStress` project, but I cannot run it:
```
It was not possible to find any compatible framework version
The framework 'Microsoft.NETCore.App', version '7.0.0' (x64) was not found.
...
```
I have `7.0.0-alpha.1.21567.1` installed globally. I vaguely remember being able to do that with 6.0...

ManickaP commented on src/libraries/System.Net.Http/tests/StressTests/HttpStress/build-local.ps1:
```diff
@@ -0,0 +1,56 @@
+    New-Item -ItemType Directory -Path $DailyDotnetRoot
+    Invoke-WebRequest -Uri https://dot.net/v1/dotnet-install.ps1 -OutFile "$DailyDotnetRoot\dotnet-install.ps1"
+    & "$DailyDotnetRoot\dotnet-install.ps1" -NoPath -Channel "$Version.1xx" -Quality daily -InstallDir $DailyDotnetRoot
+}
+else {
```
NIT: 
```suggestion
} else {
```
???

ManickaP commented on src/libraries/System.Net.Http/tests/StressTests/HttpStress/build-local.sh:
```diff
@@ -0,0 +1,64 @@
+daily_dotnet_root=./.dotnet-daily
+
+stress_configuration="Release"
+if [ "$1" != "" ]
+then
```
```suggestion
if [ -z "$1" ]; then
```

antonfirsov replied:
I need to change a bunch of scripts though, I used it consistently in this way.

ManickaP commented on src/libraries/System.Net.Security/tests/StressTests/SslStress/build-local.sh:
```diff
@@ -0,0 +1,44 @@
+#!/usr/bin/env bash
```
Is this script same (or very similar) as the one in `HttpStress`? Should we somehow share it (in eng dir potentially)?

antonfirsov replied:
HttpStress doesn't need `Microsoft.AspNetCore.App` so in that case it's much simpler. There is a lot of duplication including msbuild stuff and scripts, but moving these bits to a shared location would make things only harder to maintain. It's already pretty hard to track what lives in `/eng/docker` vs individual stress projects.

If we had a 3rd stress project that would make these bits worth to dedupe.

antonfirsov commented:
@ManickaP docker or local? In case of local, did you use `build-local.sh` and the generated runscript which uses testhost to `dotnet exec` the tests?

https://github.com/dotnet/runtime/blob/48a960215db417874803660e0667dce13df42dd5/src/libraries/System.Net.Http/tests/StressTests/HttpStress/Readme.md#running-the-suite-locally

ManickaP commented on src/libraries/System.Net.Http/tests/StressTests/HttpStress/build-local.sh:
```diff
@@ -0,0 +1,64 @@
+stress_configuration="Release"
+if [ "$1" != "" ]
+then
+    stress_configuration=${1,,}                   # Lowercase all characters in $1
+    stress_configuration=${stress_configuration^} # Uppercase first character
```
I don't think this really matter, I've been using lower-case interchangeably.

As a result, the whole if-else could be reduced to `stress_configuration="${1:-Release}"`

antonfirsov replied:
Sometimes the scripts refer to filesystem locations based on the configuration variable values:

https://github.com/dotnet/runtime/blob/48a960215db417874803660e0667dce13df42dd5/src/libraries/System.Net.Http/tests/StressTests/HttpStress/build-local.sh#L38

ManickaP replied:
I learned the hard way it does matter in certain scenarios :rofl:. 
However, I was thinking how to make this more convenient to use, e.g.: `testhost_root=$(ls -t -d $repo_root/artifacts/bin/testhost/net$version-Linux-*-x64 | head -n 1)` which picks the most recent testhost dir if you do not specify which one you want to use.

So the script would work nicely without providing any argument with either release or debug testhost builds.

ManickaP replied:
BTW, this is just a suggestion for possible future improvement. I certainly won't hold the PR on this.

antonfirsov replied:
For now I just made it fail with a verbose message if you build it with the wrong config.

ManickaP reviewed (commented):
Proper first pass through the change (it's big :smile:)
I haven't gotten to fully test this yet though.

ManickaP commented:
>  In case of local, did you use build-local.sh

No, I was able to build and run against globally installed SDK/runtime, which I've used extensively. Note that my connection is not so great to pull SDKs all the time, I usually install them globally once for all clones I have.

antonfirsov commented:
>  No, I used to be able to build and run against globally installed SDK/runtime, which I've used extensively. Note that my connection is not so great to pull SDKs all the time, I usually install them globally once for all clones I have.

After this PR, you really need the testhost to run the stress project. I don't know if there is a reliable way to hack the global `dotnet` host & runtime to make it work with the live-built bits.

Note that my script does `dotnet-install` only the first time after cloning the reop just like `build.sh` in the root. If you really prefer to grab your globally installed `Microsoft.AspNetCore.App`, you need to copy it to testhost manually.  If your main concern against the buildscript is the bandwidth usage, maybe we should expose a parameter `aspnetlocation`, so you can do:

```bash
./build-local.sh -aspnetlocation "/usr/share/dotnet/shared/Microsoft.AspNetCore.App/7.0.0-alpha.1.21567.1`
```

ManickaP commented on src/libraries/System.Net.Http/tests/StressTests/HttpStress/build-local.sh:
```diff
@@ -0,0 +1,64 @@
+    stress_configuration=${stress_configuration^} # Uppercase first character
+fi
+
+libraries_configuration="Release"
```
Maybe Debug might be better default option??? I assume that most of us build libs in Debug (only CLR in Release).

antonfirsov replied:
I always build libraries in Release when it comes to stress testing, but maybe it's just me. Doesn't a Debug build change timing characteristics significantly, meaning that you have less chance to reproduce an issue that has been detected in Release?

wfurt replied:
Preferably, one could choose. While Release is what we ship, Debug builds have more Asserts and extra checks. So it would be nice to know we can pass all of them under stress.

antonfirsov replied:
This is a parameter, so you can always choose. What we are trying to figure out is the best default for our everyday workflow when running stress tests for some reason.

ManickaP replied:
Yes, you can chose, it's just about what's the default. This looks like even split so let's keep the `Release` as default, I don't mind.

ManickaP reviewed (approved):
Thanks!
The script for running stress with your local bits is very convenient, I like it a lot!

antonfirsov committed:
nits

antonfirsov committed:
improve build-local.sh

antonfirsov committed:
improve the rest of the scripts

antonfirsov commented:
/azp run runtime-libraries stress-http

antonfirsov commented:
/azp run runtime-libraries stress-ssl
